File: model-client/src/main/java/org/modelix/model/operations/OTBranch.java
package org.modelix.model.operations;

import de.q60.mps.shadowmodels.runtime.model.persistent.IBranch;
import de.q60.mps.shadowmodels.runtime.model.persistent.IBranchListener;
import de.q60.mps.shadowmodels.runtime.model.persistent.IIdGenerator;
import de.q60.mps.shadowmodels.runtime.model.persistent.IReadTransaction;
import de.q60.mps.shadowmodels.runtime.model.persistent.ITransaction;
import de.q60.mps.shadowmodels.runtime.model.persistent.ITree;
import de.q60.mps.shadowmodels.runtime.model.persistent.IWriteTransaction;
import io.vavr.Tuple;
import io.vavr.Tuple2;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

public class OTBranch implements IBranch {

  private IBranch branch;
  private IIdGenerator idGenerator;
  private List<IAppliedOperation> operations = new ArrayList<>();
  private final Object operationsLock = new Object();

  public OTBranch(IBranch branch, IIdGenerator idGenerator) {
    this.branch = branch;
    this.idGenerator = idGenerator;
  }

  public void operationApplied(IAppliedOperation op) {
    synchronized (operationsLock) {
      operations.add(op);
    }
  }

  public List<IAppliedOperation> getNewOperations() {
    synchronized (operationsLock) {
      List<IAppliedOperation> result = operations;
      operations = new ArrayList<IAppliedOperation>();
      return result;
    }
  }

  public Tuple2<List<IAppliedOperation>, ITree> getOperationsAndTree() {
    synchronized (operationsLock) {
      return Tuple.of(getNewOperations(), computeRead(() -> getTransaction().getTree()));
    }
  }

  @Override
  public void addListener(IBranchListener l) {
    branch.addListener(l);
  }
  @Override
  public void removeListener(IBranchListener l) {
    branch.removeListener(l);
  }
  @Override
  public boolean canRead() {
    return branch.canRead();
  }
  @Override
  public boolean canWrite() {
    return branch.canWrite();
  }
  @Override
  public <T> T computeRead(Supplier<T> computable) {
    checkNotEDT();
    return branch.computeRead(computable);
  }
  @Override
  public <T> T computeWrite(Supplier<T> computable) {
    checkNotEDT();
    return branch.computeWrite(computable);
  }
  @Override
  public IReadTransaction getReadTransaction() {
    return branch.getReadTransaction();
  }
  @Override
  public ITransaction getTransaction() {
    return wrap(branch.getTransaction());
  }
  @Override
  public IWriteTransaction getWriteTransaction() {
    return wrap(branch.getWriteTransaction());
  }
  @Override
  public void runRead(Runnable runnable) {
    checkNotEDT();
    branch.runRead(runnable);
  }
  @Override
  public void runWrite(Runnable runnable) {
    checkNotEDT();
    branch.runWrite(runnable);
  }

  public ITransaction wrap(ITransaction t) {
    return (t instanceof IWriteTransaction ? wrap((IWriteTransaction) t) : t);
  }

  public IWriteTransaction wrap(IWriteTransaction t) {
    return new OTWriteTransaction(t, this, idGenerator);
  }

  protected void checkNotEDT() {
  }

}

File: model-client/src/main/kotlin/org/modelix/model/operations/OTBranch.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */

package org.modelix.model.operations

import io.vavr.Tuple
import io.vavr.Tuple2
import org.modelix.model.api.*
import java.util.*
import java.util.function.Supplier

class OTBranch(private val branch: IBranch, private val idGenerator: IIdGenerator) : IBranch {
    private var operations: MutableList<IAppliedOperation> = ArrayList()
    private val operationsLock = Any()
    fun operationApplied(op: IAppliedOperation) {
        synchronized(operationsLock) { operations.add(op) }
    }

    val newOperations: List<IAppliedOperation>
        get() {
            synchronized(operationsLock) {
                val result: List<IAppliedOperation> = operations
                operations = ArrayList()
                return result
            }
        }

    val operationsAndTree: Tuple2<List<IAppliedOperation>, ITree>
        get() {
            synchronized(operationsLock) { return Tuple.of(newOperations, computeRead<ITree>(Supplier<ITree> { transaction.tree!! })) }
        }

    override fun addListener(l: IBranchListener?) {
        branch.addListener(l)
    }

    override fun removeListener(l: IBranchListener?) {
        branch.removeListener(l)
    }

    override fun canRead(): Boolean {
        return branch.canRead()
    }

    override fun canWrite(): Boolean {
        return branch.canWrite()
    }

    override fun <T> computeRead(computable: Supplier<T>?): T {
        checkNotEDT()
        return branch.computeRead(computable)
    }

    override fun <T> computeWrite(computable: Supplier<T>?): T {
        checkNotEDT()
        return branch.computeWrite(computable)
    }

    override val readTransaction: IReadTransaction?
        get() = branch.readTransaction

    override val transaction: ITransaction
        get() = wrap(branch.transaction)

    override val writeTransaction: IWriteTransaction
        get() = wrap(branch.writeTransaction)

    override fun runRead(runnable: Runnable?) {
        checkNotEDT()
        branch.runRead(runnable)
    }

    override fun runWrite(runnable: Runnable?) {
        checkNotEDT()
        branch.runWrite(runnable)
    }

    fun wrap(t: ITransaction?): ITransaction {
        return if (t is IWriteTransaction) wrap(t as IWriteTransaction?) else t!!
    }

    fun wrap(t: IWriteTransaction?): IWriteTransaction {
        return OTWriteTransaction(t!!, this, idGenerator)
    }

    protected fun checkNotEDT() {}
}
